test: Remove the TC_XXX macro in tests/arch/arm/arm_runtime_nmi

hunterhu3000 opened the pull request:
As the test case is using ztest framework, so the TC_XXX macro
is never needed.

Signed-off-by: Hu Zhenyu <zhenyu.hu@intel.com>

stephanosio commented on tests/arch/arm/arm_runtime_nmi/src/arm_runtime_nmi.c:
```diff
@@ -22,8 +22,6 @@ static void nmi_test_isr(void)
 {
 	printk("NMI received (test_handler_isr)! Rebooting...\n");
 	/* ISR triggered correctly: test passed! */
-	TC_END_RESULT(TC_PASS);
-	TC_END_REPORT(TC_PASS);
```
But this would effectively make the `test_arm_runtime_nmi` report success even if the NMI ISR did not execute, in which case it should report a failure.

enjiamai replied:
Sorry, I will double check that. Thanks for pointing this out.

enjiamai replied:
Thanks for the reminder from @stephanosio, we updated it.

stephanosio commented:
If we are to do this, twister should at least check for `NMI received (test_handler_isr)! Rebooting...` in the console output.

enjiamai commented:
It seems like the original code have some problem (after changing to using ztest).  This test passed in all the boards. The updated changes catch the testcase failed on mps2_an521_ns.

enjiamai commented on tests/arch/arm/arm_runtime_nmi/src/arm_runtime_nmi.c:
```diff
@@ -60,6 +60,10 @@ void test_arm_runtime_nmi(void)
 	/* Trigger NMI: Should fire immediately */
 	SCB->ICSR |= SCB_ICSR_NMIPENDSET_Msk;
+
+	k_sleep(K_MSEC(1000));
```
I think the delay can be shorter here. Or, try to have no delay here. Because the NMI should react right away.

hunterhu3000 replied:
tried 10 ms, the result is the same

joerchan replied:
Delay does not matter, maybe:
```
	__ISB();
	__DSB();
```
But other than that you can just run the assert immediately.

hunterhu3000 replied:
OK, will remove the delay.

LixinGuoX commented:
Hi, @hunterhu3000 , could you try adding `filter: CONFIG_ARMV6_M_ARMV8_M_BASELINE or CONFIG_ARMV7_M_ARMV8_M_MAINLINE and not CONFIG_NULL_POINTER_EXCEPTION_DETECTION_NONE` in yaml file.
For reference: https://theotherjimmy.github.io/blog/blog/2022-01-24-bug-notes-str-alloc-copy.html

joerchan commented on tests/arch/arm/arm_runtime_nmi/src/arm_runtime_nmi.c:
```diff
@@ -18,12 +18,13 @@
+
 static void nmi_test_isr(void)
 {
 	printk("NMI received (test_handler_isr)! Rebooting...\n");
```
Can you please remove the "Rebooting..." part as well. I don't see a reboot anywhere.

joerchan reviewed (changes requested):
Test case is failing on all platforms for TF-M / Non-secure.

The issue is that SCB.AIRCR.BFHFNMINS bit is not set, so the non-secure write to SCB.ICSR.NMIPENDSET is ignored.

Maybe disable / quarantine the test-cases until the linked issue has been fixed?
Getting this change in shouldn't be dependent on the bug being fixed.

hunterhu3000 commented:
> Test case is failing on all platforms for TF-M / Non-secure.
> 
> The issue is that SCB.AIRCR.BFHFNMINS bit is not set, so the non-secure write to SCB.ICSR.NMIPENDSET is ignored.
> 
> Maybe disable / quarantine the test-cases until the linked issue has been fixed? Getting this change in shouldn't be dependent on the bug being fixed.

It seems using "and not CONFIG_NULL_POINTER_EXCEPTION_DETECTION_NONE" can also filter/skip the test case on mps2_an521_ns. Is it OK for you?

joerchan commented:
> It seems using "and not CONFIG_NULL_POINTER_EXCEPTION_DETECTION_NONE" can also filter/skip the test case on mps2_an521_ns. Is it OK for you?

I don't know how this works, would `NOT BUILD_WITH_TFM` also fix it?

I don't see the link between the null pointer exception and running the test on this board. But Ideally I think this should filter out non-secure boards in general if they are using TF-M.

@PerMac Do you know if this filter would work for disabling the nRF9160 board in our twister run in nordic boards?

joerchan commented:
@hunterhu3000 
I think the null-pointer filter is related to this:
```
config NULL_POINTER_EXCEPTION_DETECTION_NONE
	bool "No null pointer exception detection"
...
config NULL_POINTER_EXCEPTION_DETECTION_DWT
	bool "Use DWT for null pointer exception detection"
	depends on CPU_CORTEX_M_HAS_DWT
	depends on !TRUSTED_EXECUTION_NONSECURE
...
config NULL_POINTER_EXCEPTION_DETECTION_MPU
	bool "Use MPU for null pointer exception detection"
	depends on !TRUSTED_EXECUTION_NONSECURE
```

As you can see `NULL_POINTER_EXCEPTION_DETECTION_NONE` is the only allowed option for `TRUSTED_EXECUTION_NONSECURE`.
Using it as the filter is indirect. I suggest using BUILD_WITH_TFM instead, since TF-M has not allowed the non-secure application  to use the NMI interrupt.

hunterhu3000 committed:
test: Remove the TC_XXX macro in tests/arch/arm/arm_runtime_nmi

As the test case is using ztest framework, so the TC_XXX macro
is never needed.

Signed-off-by: Hu Zhenyu <zhenyu.hu@intel.com>

hunterhu3000 commented:
> 

NOT BUILD_WITH_TFM also fixes this problem. Let's use this one.
